pkg/remote/write.go

```diff
@@ -0,0 +1,232 @@
+
+func WithLogger(l log.Logger) WriteOpts {
+	return func(w *Write) {
+		w.logger = l
+	}
+}
+
+func NewWrite(opts ...WriteOpts) *Write {
+	w := &Write{
+		logger: log.NewNopLogger(),
+
+		samplePerTenantID:   make(map[string][]pulsar.ReceivedSample),
+		deadlinePerTenantID: make(map[string]time.Time),
+
+		BatchSize:     100,
+		BatchMaxDelay: 5 * time.Second,
+	}
+
+	for _, opt := range opts {
+		opt(w)
+	}
+
+	return w
+}
+
+type headerRoundtripper struct {
+	upstream http.RoundTripper
+}
+
+func (t *headerRoundtripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	// if tenant ID is set, expose it in the header
+	tenantID := mcontext.TenantIDFromContext(req.Context())
+	if tenantID != "" {
+		req.Header.Set(mcontext.HTTPHeaderTenantID, tenantID)
+	}
+
+	// overwrite the user agent header
+	req.Header.Set("User-Agent", fmt.Sprintf("%s/%s", version.AppName(), version.AppVersion()))
+
+	return t.upstream.RoundTrip(req)
+}
+
+type ClientConfig = remote.ClientConfig
+
+func NewWriteClient(conf *ClientConfig) (remote.WriteClient, error) {
+	clientInterface, err := remote.NewWriteClient("remote-write", conf)
+	if err != nil {
+		return nil, err
+	}
+
+	client, ok := clientInterface.(*remote.Client)
+	if !ok {
+		return nil, fmt.Errorf("client of unexpected type %T", clientInterface)
+	}
+
+	// add custom rountripper to modify user-agent and tenant header if necessary
+	client.Client.Transport = &headerRoundtripper{upstream: client.Client.Transport}
+
+	return client, nil
+}
+
+func (w *Write) Run(ctx context.Context, sampleCh chan pulsar.ReceivedSample, client remote.WriteClient) error {
+	tick := time.NewTicker(100 * time.Millisecond)
+	defer tick.Stop()
+
+	// this is set when a retry able error happend
+	errRemoteWriteRetryable := false
+	blockingSampleCh := make(chan pulsar.ReceivedSample)
+
+	sample := func() chan pulsar.ReceivedSample {
+		// don't return new samples if err has happend
+		if errRemoteWriteRetryable {
+			return blockingSampleCh
+		}
+		return sampleCh
+	}
+
+receive:
+	for {
+		// wait for either tick or incoming sample, handle closed context
+		select {
+		case <-ctx.Done():
+			break receive
+		case <-tick.C:
+			// do nothing
+		case sample := <-sample():
+			tenantID := mcontext.TenantIDFromContext(sample.Context)
+			if _, ok := w.samplePerTenantID[tenantID]; !ok {
+				w.samplePerTenantID[tenantID] = []pulsar.ReceivedSample{}
+				w.deadlinePerTenantID[tenantID] = time.Now().Add(w.BatchMaxDelay)
+			}
+
+			w.samplePerTenantID[tenantID] = append(w.samplePerTenantID[tenantID],
+				sample,
+			)
+		}
+
+		// reset retryable error
+		errRemoteWriteRetryable = false
+
+		// loop through tenants and find metrics to send
+		for tenantID, samples := range w.samplePerTenantID {
+			logger := tenantIDLogger(w.logger, tenantID)
+
+			batchSizeReached := len(samples) >= w.BatchSize
+			pastDeadline := w.deadlinePerTenantID[tenantID].Before(time.Now())
+
+			if !batchSizeReached && !pastDeadline {
+				continue
```

Comment:
I'm not sure if I completely understand what's going on here, correct me if I'm wrong. 

Every time a sample comes in through the `chan` returned by `sample()` this loop is going to iterate over all tenants to check whether the deadline for a tenant has exceeded or a client's batch has reached the max batch size, right? That seems like quite a lot of overhead in situations where the rate of samples is high, maybe it would make sense to only do these checks after every 100 samples or 1sec, whichever happens first, or something like that.

Reply:
Good point! That makes a lot of sense. 

When I benchmark them against each other, you can see it fairly clearly:

Before:
```
BenchmarkWriter1Tenants100BatchSize-8             812914              1309 ns/op            812914 samples
BenchmarkWriter50Tenants100BatchSize-8              1599            662711 ns/op             79950 samples
BenchmarkWriter100Tenants100BatchSize-8              450           2801191 ns/op             45000 samples
BenchmarkWriter500Tenants100BatchSize-8               33          62741805 ns/op             16500 samples
```

With a583d305cdc5b1c9f7cbd5822682b67c9a1b373c:
```
BenchmarkWriter1Tenants100BatchSize-8            1075980              1041 ns/op           1075980 samples
BenchmarkWriter50Tenants100BatchSize-8             18751             62577 ns/op            937550 samples
BenchmarkWriter100Tenants100BatchSize-8            10000            137519 ns/op           1000000 samples
BenchmarkWriter500Tenants100BatchSize-8              913           1354543 ns/op            456500 samples
```